### PR TITLE
Load from oldest snapshot if specified version is null in request.

### LINKED ISSUE #4049: Fix Loading From Ops
## Describe the bug
With detached container we always need to load from a snapshot to start. Currently if the container request passes `headers.version === null` we skip all snapshots, and load from ops only. This now causes a problem. We should update to load from the oldest snapshot when `headers.version === null`

https://github.com/microsoft/FluidFramework/blob/2b21b9dc5ab906a07924418e01d15bbf2d586237/packages/loader/container-loader/src/container.ts#L205
https://github.com/microsoft/FluidFramework/blob/2b21b9dc5ab906a07924418e01d15bbf2d586237/packages/loader/container-loader/src/container.ts#L1000

### PR DESCRIPTION
**Pull request by jatgarg**
Fixes: https://github.com/microsoft/FluidFramework/issues/4049

### CODE REVIEW

File: packages/loader/container-loader/src/container.ts [hunk 1]
```diff
@@ -1614,17 +1613,21 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
...
 
         if (version !== undefined) {
             this._loadedFromVersion = version;
             return await this.storageService.getSnapshotTree(version) ?? undefined;
         } else if (specifiedVersion !== undefined) {
             // We should have a defined version to load from if specified version requested
-            this.logger.sendErrorEvent({ eventName: "NoVersionFoundWhenSpecified", specifiedVersion });
+            this.logger.sendErrorEvent({
+                eventName: "NoVersionFoundWhenSpecified",
+                specifiedVersion: specifiedVersion ?? "Oldest",
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/container.ts:1629]: this is wrong, as it looses the distinction between null and undefined. i would just log specified version directly

File: packages/drivers/routerlicious-driver/src/documentStorageService.ts [hunk 2]
```diff
@@ -49,8 +49,23 @@ export class DocumentStorageService implements IDocumentStorageService {
...
+    public async getVersions(versionId: string | null, count: number): Promise<IVersion[]> {
+        // If version is null, that means we need to fetch the oldest snapshot. For that use a fairly
+        // high number for the number of snapshot versions to be fetched. The last snapshot version would
+        // be the oldest. This is required as we don't have any way to refer to oldest snapshot for now.
+        // TODO: Think of api to fetch oldest snapshot(server/client side) like fetch snapshot with seq number 0.
+        const commits = await this.manager.getCommits(
+            versionId ? versionId : this.id,
+            versionId !== null ? count : 100000,
+        );
+        if (versionId === null) {
```
Comment (anthony-murphy) [inline packages/drivers/routerlicious-driver/src/documentStorageService.ts:61]: is it possible to write an test for this?

**Comment by anthony-murphy** [comment]
also. i like the branch name 😀

File: packages/loader/container-loader/src/container.ts (hunk 1 above)
Comment (jatgarg) [reply packages/loader/container-loader/src/container.ts:1629]: Undefined is already eliminated in if condition. Can add an explicit check.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 3]
```diff
@@ -473,7 +473,12 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
 
         return getWithRetryForTokenRefresh(async (options) => {
             const storageToken = await this.getStorageToken(options, "GetVersions");
-            const { url, headers } = getUrlAndHeadersWithAuth(`${this.snapshotUrl}/versions?count=${count}`, storageToken);
+            // If version is null, that means we need to fetch the oldest snapshot. For that use 50 for the number of
+            // snapshot versions to be fetched. The last snapshot version would be the oldest. ODSP doesn't store more
+            // than 50 snapshots and always retain the oldest snapshot. Also they don't allow requesting more than 50
+            // snapshots. This is required as we don't have any way to refer to oldest snapshot for now.
+            // TODO: Think of api to fetch oldest snapshot(server/client side) like fetch snapshot with seq number 0.
+            const { url, headers } = getUrlAndHeadersWithAuth(`${this.snapshotUrl}/versions?count=${blobid !== null ? count : 50}`, storageToken);
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:481]: I'm confused on 50. Where does it come from? I.e. why 50, not count? Caller is always in control of the count, no?

File: packages/loader/container-loader/src/loader.ts [hunk 4]
```diff
@@ -433,7 +433,7 @@ export class Loader extends EventEmitter implements ILoader {
         // If set in both query string and headers, use query string
         request.headers[LoaderHeader.version] = parsed.version ?? request.headers[LoaderHeader.version];
 
-        // Version === null means not use any snapshot.
+        // Version === null means use oldest snapshot.
```
Comment (vladsud) [inline packages/loader/container-loader/src/loader.ts:436]: Can you please clarify what "oldest snapshot" means? I.e. how is it different from not passing LoaderHeader.version?
If it's the same, then I'd argue we should not allow / support null here, there is already a way to load latest - by doing nothing (not passing anything).

**Comment by vladsud** [comment]
It's a bit hard to follow this PR as it's not clear what are the semantic changes across layers and why.
For example, passing getVersions(docId) was returning latest snapshot. Are we disallowing this pattern in favor of null? If not, then do we have two ways of encoding latest? Also why null, why not undefined? undefined is more established for these thigs, and it maps really well to being default arg.
Any chance you can update comments on all impacted interfaces on what is the assumptions now for this arg?

File: packages/loader/container-loader/src/container.ts (hunk 1 above)
Comment (anthony-murphy) [reply packages/loader/container-loader/src/container.ts:1629]: regardless, it's best to log the actual value to explicitly understand what's happening in the code. someone could change something and then i need to correlate code version to understand what the original value was.

File: packages/loader/container-loader/src/loader.ts (hunk 4 above)
Comment (anthony-murphy) [reply packages/loader/container-loader/src/loader.ts:436]: i believe undefined in latest, null is oldest, and anything else is specific

Comment (anthony-murphy) [reply packages/loader/container-loader/src/loader.ts:436]: null used to mean load from ops, but that doesn't really exist for detached anymore. it would be nice if there was an op that told us to go load the summary if we haven't already, rather than doing this trickery with storage

**Comment by vladsud** [comment]
Ah, got it. For some reason I read latest where it was mentioned oldest.
Can we instead of using undefined & null in the code use symbolic names for them?
I.e. 
const LatestSnapshot = undefined;
const EarliestSnapshot = null;

I'd go even further and use strings (i.e. some hardcoded names), but that always runs into chance of collisions :(

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts (hunk 3 above)
Comment (anthony-murphy) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:481]: I made this comment below, but i think using an op here might be a better solution. Basically we load from ops when null just like now, but the first op basically says you better have this summary or equivalent state to move forward from here. I like this as well as if in the future we truncate ops, we'd truncate to some summary op, and the same logic would still work.

Comment (anthony-murphy) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:481]: we might still want this at backcompat/short term, basically if we didn't load from summary, and the first op isn't summary or code proposal we need to go load some summary

File: packages/drivers/routerlicious-driver/src/documentStorageService.ts (hunk 2 above)
Comment (jatgarg) [reply packages/drivers/routerlicious-driver/src/documentStorageService.ts:61]: Yes, I tested it with both r11s and odsp.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts (hunk 3 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:481]: I think we can think of op implementation in near future as that would need server support too. They need to send that op and for that they should have the oldest snapshot version that they would send in the op. 
One other alternative is that to have a api in storage which allows fetchng snapshots or versions according to seq number. So in order to fetch the oldest snapshot we can fetch the snapshot with seq number 0. 
@tanviraumi Yes 50 is the max odsp allows in the query that we make, however they don't even store more than 10 now. As more snapshots keep coming in, they start purging snapshots, however they never purge the first one.

File: packages/loader/container-loader/src/loader.ts (hunk 4 above)
Comment (jatgarg) [reply packages/loader/container-loader/src/loader.ts:436]: https://github.com/microsoft/FluidFramework/pull/4259#discussion_r518927846

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts (hunk 3 above)
Comment (anthony-murphy) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:481]: maybe. couldn't we just stick and op in the op stream on attach as the first op

Comment (anthony-murphy) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:481]: regardless. if that is a path we think we want to go down. we should consider it for this change. so null would still mean load from ops, but we would look at the initial ops to determine if we need to load a snapshot, which is quite different than what we have here

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:481]: But I am bit skeptical about going on op path for loading for oldest snapshot because:
1.) We don't do such thing when we load from latest/specified snapshot.
2.) It would mean that we would use kind of null runtime and then look at ops and then switch context to actual runtime after fetching that snapshot. I don't think we want to do that given we moved to detached flow with one reason to get rid of it.

Comment (anthony-murphy) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:481]: 1.) And i don't think we need to i see the link between the op and the summary and bi-directional, so you can start with the summary, and then you see the op that says you must have state equivalent to some sequence to proceed, and we would know that we do. for the null case, we would see the op and realize we didn't and have to load it.

2) It's true, but i don't think we will be able to get rid of it for the forceable future anyway. I also don't have a problem with it as an internal implementation detail, which this would allow.

File: packages/loader/container-loader/src/container.ts (hunk 1 above)
Comment (arinwt) [reply packages/loader/container-loader/src/container.ts:1629]: > I'd go even further and use strings (i.e. some hardcoded names), but that always runs into chance of collisions :(

At least minimize the area in code where this distinction exists by immediately wrapping in an object or a union type including symbols or something.

**Comment by arinwt** [comment]
> I'd go even further and use strings (i.e. some hardcoded names), but that always runs into chance of collisions :(

At least minimize the area in code where this distinction exists by immediately wrapping in an object or a union type including symbols or something.

**Comment by tylerbutler** [comment]
@jatgarg Is this PR still relevant?

**Comment by jatgarg** [comment]
@tylerbutler yes, I would take it up.

**Comment by jatgarg** [comment]
Closing as no longer required.
